SeriesCruncher/app/src/main/java/com/balazs_csernai/seriescruncher/seriesdetails/ui/EpisodeAdapter.java

```diff
@@ -106,4 +108,74 @@ public SeasonsViewHolder(View view) {
             ButterKnife.inject(this, view);
         }
     }
-}
+
+    private EpisodeListItemModel getItem(int position) {
+        return items.get(mapVisiblePositionToInnerPosition(position));
+    }
+
+    private void onItemTap(int position) {
+        EpisodeListItemModel item = items.get(position);
+
+        if (!item.isEpisode()) {
+            int nextSeasonOrEndPosition = getNextSeasonOrEndPosition(position);
+
+            for (int itemPosition = position; itemPosition < nextSeasonOrEndPosition; itemPosition++) {
+                items.get(itemPosition).toggle();
+            }
+
+            int firstInsertedOrRemovedItemPosition = position + 1;
+            int modifiedItemCount = nextSeasonOrEndPosition - firstInsertedOrRemovedItemPosition;
+            int firstInsertedOrRemovedVisibleItemPosition = mapInnerPositionToVisiblePosition(firstInsertedOrRemovedItemPosition);
+
+            notifyAdapter(firstInsertedOrRemovedVisibleItemPosition, modifiedItemCount, item.isExpanded());
+        }
+    }
+
+    private void notifyAdapter(int position, int count, boolean insertion) {
+        if (insertion) {
+            notifyItemRangeInserted(position, count);
+        } else {
+            notifyItemRangeRemoved(position, count);
+        }
+    }
+
+    private int mapVisiblePositionToInnerPosition(int position) {
+        return countItemsBeforeVisiblePosition(position);
+    }
+
+    private int mapInnerPositionToVisiblePosition(int position) {
+        return position - countInvisibleItemsBeforePosition(position);
+    }
+
+    private int countItemsBeforeVisiblePosition(int position) {
+        int itemCount = 0;
+        int visibleItemCount = 0;
+        while (visibleItemCount != position && itemCount < items.size()) {
+            visibleItemCount += isVisible(items.get(++itemCount)) ? 1 : 0;
```

Comment:
In my opinion this compact solution reduces the readability of the code (although its pretty cool), plus if the isVisible method returns false, we are performing an addition with 0, which is unnecessary.
Please refactor using an if statement.